src/Hook.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Roave\YouAreUsingItWrong;
+
+use Composer\Composer;
+use Composer\EventDispatcher\EventSubscriberInterface;
+use Composer\IO\IOInterface;
+use Composer\Plugin\PluginInterface;
+use Composer\Script\Event;
+use Composer\Script\ScriptEvents;
+use PackageVersions\Versions;
+use Psalm\Internal\Analyzer\ProjectAnalyzer;
+use Psalm\Internal\Provider\FileProvider;
+use Psalm\Internal\Provider\Providers;
+use Psalm\IssueBuffer;
+use Roave\YouAreUsingItWrong\Composer\Project;
+use Roave\YouAreUsingItWrong\Psalm\Configuration;
+use Roave\YouAreUsingItWrong\Psalm\ProjectFilesToBeTypeChecked;
+use RuntimeException;
+use function define;
+use function defined;
+use function getcwd;
+use function microtime;
+
+/** @internal this is a composer plugin: do not rely on it in your sources */
+final class Hook implements PluginInterface, EventSubscriberInterface
+{
+    private const THIS_PACKAGE_NAME = 'roave/you-are-using-it-wrong';
+
+    /**
+     * {@inheritDoc}
+     */
+    public function activate(Composer $composer, IOInterface $io) : void
+    {
+        // Nothing to do here, as all features are provided through event listeners
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public static function getSubscribedEvents() : array
+    {
+        return [
+            ScriptEvents::POST_INSTALL_CMD => 'runTypeChecks',
+            ScriptEvents::POST_UPDATE_CMD  => 'runTypeChecks',
+        ];
+    }
+
+    /**
+     * @throws RuntimeException
+     */
+    public static function runTypeChecks(Event $composerEvent) : void
+    {
+        $io       = $composerEvent->getIO();
+        $composer = $composerEvent->getComposer();
+        $project  = Project::fromComposerInstallationContext($composer->getPackage(), $composer->getLocker(), getcwd());
+
+        if (! $project->strictTypeChecksAreEnforcedByLocalInstallation()) {
+            $io->write('<info>' . self::THIS_PACKAGE_NAME . ':</info> plugin not installed locally - skipping type checks...');
+
+            return;
+        }
+
+        if ($project->alreadyHasOwnPsalmConfiguration()) {
+            $io->write('<info>' . self::THIS_PACKAGE_NAME . ':</info> psalm.xml detected - assuming static analysis will run later; not running psalm now');
+
+            return;
+        }
+
+        $io->write('<info>' . self::THIS_PACKAGE_NAME . ':</info> checking strictly type-checked packages...');
+
+        self::analyseProject($project);
+
+        $io->write('<info>' . self::THIS_PACKAGE_NAME . ':</info> ... done checking strictly type-checked packages');
+    }
+
+    private static function analyseProject(Project $project) : void
+    {
+        defined('PSALM_VERSION') || define('PSALM_VERSION', Versions::getVersion('vimeo/psalm'));
```

Comment:
Is there a reason to define these constants like this, instead of within an if statement?

Reply:
Not really, this is just a very common pattern for constant declaration (coming from older ages, where non-namespaced constants were still a thing). Can surely use `if` instead

Reply:
I never really worked with `defined`and `define`, so this syntax is somewhat new for me. If it is the 'standard' for defining these then i'd leave it like that.

Reply:
I will change it: no need to add another syntax for `if` - just harder to read

Reply:
PSR-2 requires three lines for an if statement, so this way is a lot more compact, although how it works is less obvious. Relies on the short-circuit evaluation of ||

Reply:
I don't mind longer syntax here: the computed complexity is the same